espnet2/bin/enh_inference_streaming.py

```diff
@@ -0,0 +1,545 @@
+    ngpu: int,
+    seed: int,
+    num_workers: int,
+    log_level: Union[int, str],
+    data_path_and_name_and_type: Sequence[Tuple[str, str, str]],
+    key_file: Optional[str],
+    train_config: Optional[str],
+    model_file: Optional[str],
+    model_tag: Optional[str],
+    inference_config: Optional[str],
+    allow_variable_data_keys: bool,
+    ref_channel: Optional[int],
+    enh_s2t_task: bool,
+):
+    assert check_argument_types()
+    if batch_size > 1:
+        raise NotImplementedError("batch decoding is not implemented")
+    if ngpu > 1:
+        raise NotImplementedError("only single GPU decoding is supported")
+
+    logging.basicConfig(
+        level=log_level,
+        format="%(asctime)s (%(module)s:%(lineno)d) %(levelname)s: %(message)s",
+    )
+
+    if ngpu >= 1:
+        device = "cuda"
+    else:
+        device = "cpu"
+
+    # 1. Set random-seed
+    set_all_random_seed(seed)
+
+    # 2. Build separate_speech
+    separate_speech_kwargs = dict(
+        train_config=train_config,
+        model_file=model_file,
+        inference_config=inference_config,
+        ref_channel=ref_channel,
+        device=device,
+        dtype=dtype,
+        enh_s2t_task=enh_s2t_task,
+    )
+    separate_speech = SeparateSpeechStreaming.from_pretrained(
+        model_tag=model_tag,
+        **separate_speech_kwargs,
+    )
+
+    # get frame_size and hop_size of the streaming model
+    frame_size, hop_size = separate_speech.frame_size, separate_speech.hop_size
+
+    # 3. Build data-iterator
+    loader = EnhancementTask.build_streaming_iterator(
+        data_path_and_name_and_type,
+        dtype=dtype,
+        batch_size=batch_size,
+        key_file=key_file,
+        num_workers=num_workers,
+        preprocess_fn=EnhancementTask.build_preprocess_fn(
+            separate_speech.enh_train_args, False
+        ),
+        collate_fn=EnhancementTask.build_collate_fn(
+            separate_speech.enh_train_args, False
+        ),
+        allow_variable_data_keys=allow_variable_data_keys,
+        inference=True,
+    )
+
+    # 4. Start dataset for-loop
+    output_dir = Path(output_dir).expanduser().resolve()
+    writers = []
+    for i in range(separate_speech.num_spk):
+        writers.append(
+            SoundScpWriter(f"{output_dir}/wavs/{i + 1}", f"{output_dir}/spk{i + 1}.scp")
+        )
+
+    import tqdm
+
+    for i, (keys, batch) in tqdm.tqdm(enumerate(loader)):
+        logging.info(f"[{i}] Enhancing {keys}")
+        assert isinstance(batch, dict), type(batch)
+        assert all(isinstance(s, str) for s in keys), keys
+        _bs = len(next(iter(batch.values())))
+        assert len(keys) == _bs, f"{len(keys)} != {_bs}"
+        batch = {k: v for k, v in batch.items() if not k.endswith("_lengths")}
+
+        speech = batch["speech_mix"]
+
+        # split continuous speech into small chunks to simulate streaming
+        speech_sim_chunks, rest_pad = split_audio(speech, frame_size, hop_size)
```

Comment:
We need to be careful when manually generating the frames because it should be consistent with the procedure defined in the encoder. For example, depending on the padding mode of the encoder, we may need to apply some paddings (can be non-zero) to the left/right side of the original input signal.

Reply:
If we do that manually, then we should ignore the builtin padding operations in the actually encoder module. Things become too complicated...

Reply:
Still, keep it manually outside the model.  Builtin padding is ignored by new `forward_streaming` in the encoder and decoder.